troop/views.py

```diff
@@ -106,3 +131,81 @@ def form_invalid(self, form):
         response = super().form_invalid(form)
         response.status_code = 422  # Unprocessable Entity
         return response
+
+
+class NamiSearchView(OnlyTroopManagerMixin, generic.FormView):
+    form_class = NamiSearchForm
+    template_name = "troop/nami_search_form.html"
+
+    def nami(self):
+        if not settings.NAMI_USERNAME or not settings.NAMI_PASSWORD:
+            raise Exception("empty nami settings")
+
+        return Nami(
+            {
+                "username": settings.NAMI_USERNAME,
+                "password": settings.NAMI_PASSWORD,
+                "session_file": settings.NAMI_SESSION,
+            }
+        )
+
+    def form_valid(self, form):
+        kwargs = {"troop": self.request.troop.number}
+        data = {"nami": form.cleaned_data["nami"]}
+
+        try:
+            p = Participant.objects.filter(**data).get()
+
+            if p.troop_id != self.request.troop.pk:
+                messages.add_message(
+                    self.request,
+                    messages.ERROR,
+                    _("Participant already registered with another troop."),
+                )
+
+                response = super().form_invalid(form)
+                response.status_code = 422  # Unprocessable Entity
+                return response
+
+            messages.add_message(
+                self.request, messages.INFO, _("Participant already registered."),
+            )
+
+            kwargs["pk"] = p.id
+            self.success_url = reverse("troop:participant.edit", kwargs=kwargs)
+            return super().form_valid(form)
+        except Participant.DoesNotExist:
```

Comment:
This is a huge try-catch-block although it only depends on the first line, right? I think we should really limit the block the try-except to the first line. Alternatively, we could do ` if Participant.[...].exists():`

Reply:
I refactored it to make it clearer :-)

Reply:
This is better but `self._redirect_to_participant(p, form)` won't throw `Participant.DoesNotExist`, right? So I'd put it outside of the block. Makes it way easier to check where an exception happened if we keep all try-catch blocks small.

Reply:
I don't understand how you expect this to be solved. If don't want to call `_redirect_to_participant` in case of an exception, that's why I let it inside the try.

Following your suggestion, it would be:
```
        try:
            p = Participant.objects.filter(**data).get()
        except Participant.DoesNotExist:
            pass
        return self._redirect_to_participant(p, form)
```
Which means that the code after would never be called (and it will fail, since `p` will not be defined in case of an exception).

Please show me a code example of what you mean.

Reply:
```
p = None

try:
    p = Participant.objects.filter(**data).get()
except Participant.DoesNotExist:
    pass

if p:
    return self._redirect_to_participant(p, form)
```

Alternative version:

```
if Participant.objects.filter(**data).exists():
    p = Participant.objects.filter(**data).get()
    return self._redirect_to_participant(p, form)
```

Reply:
I don't find this easier to read, but I have changed it to comply with your request (and it is still better than your second version, which makes 2 database requests and is not [EAFP](https://docs.python.org/3/glossary.html#EAFP)-like )